From 84ebadcd706e546c4863bff063389a82590fe597 Mon Sep 17 00:00:00 2001
From: Mateusz Guzik <mguzik@redhat.com>
Date: Wed, 25 Jun 2014 20:22:43 -0400
Subject: [ipc] fix compat msgrcv with negative msgtyp

Message-id: <20140625202242.GB21977@mguzik.redhat.com>
Patchwork-id: 84832
O-Subject: [RHEL7 PATCH] ipc: fix compat msgrcv with negative msgtyp
Bugzilla: 1042807
RH-Acked-by: Jarod Wilson <jarod@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1042807

Upstream:
commit e7ca2552369c1dfe0216c626baf82c3d83ec36bb
Author: Mateusz Guzik <mguzik@redhat.com>
Date:   Mon Jan 27 17:07:11 2014 -0800

    ipc: fix compat msgrcv with negative msgtyp

    Compat function takes msgtyp argument as u32 and passes it down to
    do_msgrcv which results in casting to long, thus the sign is lost and we
    get a big positive number instead.

    Cast the argument to signed type before passing it down.

Signed-off-by: Mateusz Guzik <mguzik@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/ipc/compat.c b/ipc/compat.c
index 892f658..d3b3760 100644
--- a/ipc/compat.c
+++ b/ipc/compat.c
@@ -381,7 +381,7 @@ COMPAT_SYSCALL_DEFINE6(ipc, u32, call, int, first, int, second,
    uptr = compat_ptr(ipck.msgp);
    fifth = ipck.msgtyp;
   }
-  return do_msgrcv(first, uptr, second, fifth, third,
+  return do_msgrcv(first, uptr, second, (s32)fifth, third,
      compat_do_msg_fill);
  }
  case MSGGET:
-- 
1.7.1